doc/source/rllib/rllib-rlmodule.rst

```diff
@@ -0,0 +1,571 @@
+            def input_specs_exploration(self) -> SpecType:
+                # Enforce that input nested dict to exploration method has a key "obs"
+                # and within that key, it has a key "global" and "local". There should 
+                # also be a key "action_mask"
+                return [("obs", "global"), ("obs", "local"), "action_mask"]
+
+
+
+.. tabbed:: TensorShape Spec
+
+    .. code-block:: python
+
+        from ray.rllib.models.specs.typing import SpecType
+        from ray.rllib.models.specs.torch_spec import TorchTensorSpec
+        
+        class DiscreteBCTorchModule(TorchRLModule):
+            ...
+            def input_specs_exploration(self) -> SpecType:
+                # Enforce that input nested dict to exploration method has a key "obs"
+                # and its value is a torch.Tensor with shape (b, h) where b is the
+                # batch size (determined at run-time) and h is the hidden size 
+                # (fixed at 10).
+                return {"obs": TorchTensorSpec("b, h", h=10)}
+
+
+
+.. tabbed:: Type Spec
+
+    .. code-block:: python
+
+        from ray.rllib.models.specs.typing import SpecType
+        import torch
+        
+        class DiscreteBCTorchModule(TorchRLModule):
+            ...
+            def output_specs_exploration(self) -> SpecType:
+                # Enforce that output nested dict from exploration method has a key
+                # "action_dist" and its value is a torch.distribution.Categorical
+                return {"action_dist": torch.distribution.Categorical}
+
+:py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` has two methods for each forward method, totaling 6 methods that can be override to describe the specs of the input and output of each method:
+
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_inference`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_inference`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_exploration`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_exploration`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_train`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_train`
+
+To learn more about this see the documentation of `SpecType`.
+
+
+
+Writing Custom Multi Agent RL Modules (Advanced)
+------------------------------------------------
+
+For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`. It is dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, plus possibly some shared modules. The base-class implementation works for most of use-cases where we need to defined independent neural for sub-groups of agents. For more complex, multi-agent use-cases, where the agents share some part of their neural network, one should inherit from this class and override the default implementation.
```

Comment:
```suggestion
For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`, which is a dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, plus possibly some shared modules. The base-class implementation works for most of use-cases where we need to defined independent neural for sub-groups of agents. For more complex, multi-agent use-cases, where the agents share some part of their neural network, one should inherit from this class and override the default implementation.
```
```suggestion
For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`. It is dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, and possibly some shared modules. The base-class implementation works for most use cases that need to define independent neural for sub-groups of agents. For more complex, multi-agent use cases, where the agents share some part of their neural network, inherit from this class and override the default implementation.
```

Reply:
I wasn't sure who "one" was. Is that the user? or the the agent? I took a guess that it's the user.